File: Leetcode/Python/_4.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # Complexity O(n + m)
        m = len(nums1) + len(nums2)
        if m % 2 == 1:
            left = right = m // 2
        else:
            left, right = m // 2 - 1, m // 2
        l1, l2 = 0, 0
        output = 0
        while l1 < len(nums1) and l2 < len(nums2):
            if l1 < len(nums1) and nums1[l1] < nums2[l2]:
                if (l1 + l2) == left:
                    output += nums1[l1]
                elif (l1 + l2) == right:
                    output += nums1[l1]
                l1 += 1
            else:
                if (l1 + l2) == left:
                    output += nums2[l2]
                elif (l1 + l2) == right:
                    output += nums2[l2]
                l2 += 1
        while l1 < len(nums1):
            if (l1 + l2) == left:
                output += nums1[l1]
            elif (l1 + l2) == right:
                output += nums1[l1]
            l1 += 1
        while l2 < len(nums2):
            if (l1 + l2) == left:
                output += nums2[l2]
            elif (l1 + l2) == right:
                output += nums2[l2]
            l2 += 1
        
        return output / 2 if left != right else output
        

class Solution:
    def findMedianSortedArraysBrute(self, nums1: List[int], nums2: List[int]) -> float:
        """Merge two array, sort it and jsut find the median of a sorted array
        Complexity O((m + n) log(m + n))
        """
        merge = nums1 + nums2
        merge = sorted(merge)
        if len(merge) % 2 != 0:
            return merge[len(merge) // 2]
        else:
            return (merge[len(merge) // 2] + merge[len(merge) // 2 - 1])/2

                         

File: Leetcode/cpp/4.cpp
#include <iostream>
#include <vector>
#include <cassert>

class Solution {
public:
    int p1 = 0, p2 = 0;

    int getMin(std::vector<int>& nums1, std::vector<int>& nums2) {
        int l1 = static_cast<int>(nums1.size());
        int l2 = static_cast<int>(nums2.size());
        
        if (p1 < l1 && p2 < l2) {
            return nums1[p1] < nums2[p2] ? nums1[p1++] : nums2[p2++];
        } else if (p1 < l1) {
            return nums1[p1++];
        } else if (p2 < l2) {
            return nums2[p2++];
        }
        return -1;
    }

    double findMedianSortedArrays(std::vector<int>& nums1, std::vector<int>& nums2) {
        int m = static_cast<int>(nums1.size()), n = static_cast<int>(nums2.size());
        
        if ((m + n) % 2 == 0) {
            for (int i = 0; i < (m + n) / 2 - 1; ++i) {
                int _ = getMin(nums1, nums2);
            }
            return (double)(getMin(nums1, nums2) + getMin(nums1, nums2)) / 2.0;

        } else {
            for (int i = 0; i < (m + n) / 2; ++i) {
                int _ = getMin(nums1, nums2);
            }
            return getMin(nums1, nums2);
        }

        return -1;
    }
};


int main() {
    Solution sol;
    std::vector<int> nums1 = {1,3}, nums2 = {2};
    double expected_output = 2.0;

    double output = sol.findMedianSortedArrays(nums1, nums2);
    assert(output == expected_output && "Test #1 failed!");

    std::cout << "Tests are passed!\n";
    return 0;
}
